File: dft.cpp
#include <bits/stdc++.h>
using namespace std;

const float pi = 3.14159265358979;

vector<complex<double>> dft(vector<complex<double>> x){
    int N = x.size();
    vector<complex<double>> X;
    for (int k = 0; k < N; k++){
        double re = 0;
        double im = 0;        
        for (int n = 0; n < N; n++){
            double phi = (2 * pi * k * n) / N;
            re += x[n].real() * cos(phi);
            im -= x[n].imag() * sin(phi);
        }
        complex<double> result (re, im);
        X.push_back(result);
    }
    return X;
}

int main(){
    string temp;
    vector<complex<double>> x, y;
    vector<complex<double>> result_x, result_y;
    ifstream x_data, y_data;
    x_data.open("/home/supperman/Documents/linhtinh/x.txt", ios::in);
    y_data.open("/home/supperman/Documents/linhtinh/y.txt", ios::in);
    ofstream res_x, res_y;
    res_x.open("/home/supperman/Documents/linhtinh/(Python & C++) Fourier Transform and Epicycle/result_x.txt", ios::out | ios::app);
    res_y.open("/home/supperman/Documents/linhtinh/(Python & C++) Fourier Transform and Epicycle/result_y.txt", ios::out | ios::app); 
    while(x_data >> temp){
        float temp_0 = stof(temp);
        complex<double> adder (temp_0, 0);
        cout << adder.real() << ' ' << adder.imag() << endl;
        x.push_back(adder); 
    }
    while (y_data >> temp){
        float temp_0 = stof(temp);
        complex<double> adder (temp_0, 0);
        cout << adder.real() << ' ' << adder.imag() << endl;
        y.push_back(adder);
    }
    result_x = dft(x);
    result_y = dft(y);
    for (unsigned int k = 0; k < result_x.size() - 1; k++){
        int freq = k;
        float amp_x = abs(result_x.at(k));
        float phase_x = arg(result_x.at(k));
        float amp_y = abs(result_y.at(k));
        float phase_y = arg(result_y.at(k));

        res_x << "[" << freq << ", " << amp_x << ", " << phase_x << "]" << endl;
        cout << result_x.at(k) << endl;

        res_y << "[" << freq << ", " << amp_y << ", " << phase_y << "]" << endl;
        cout << result_y.at(k) << endl;

    }
    x_data.close();
    y_data.close();
    res_x.close();
    res_y.close();
    cout << "DONE!" << endl;
    return 0;
}

File: dft.py
import math

x = []
y = []

def dft(re_list: list, im_list: list) -> list:
    '''
    Input: Two lists, which contains real and imagine component (both as float) of inputs.
    Output: A list of lists, which contains three elements: frequency, amplitude and phase of each epicycle.
    '''
    X = []
    N = len(x)
    k = 0
    n = 0
    
    for k in range(N):
        re = 0
        im = 0
        
        for n in range(N):
            phi = ((2 * math.pi) / N) * k * n
            re += re_list[n] * math.cos(phi) + im_list[n] * math.sin(phi)
            im -= re_list[n] * math.sin(phi) - im_list[n] * math.cos(phi)
        re = re / N
        im = im / N
        freq = k
        amp = math.sqrt(re * re + im * im)
        phase = math.atan2(im, re)
        result = [freq, amp, phase]
        X.append(result)
    return X

with open('./x.txt', 'r') as x_data:
    x = x_data.read().split('\n')
    temp = [float(i) for i in x if i != '']
    x = temp
    

with open('./y.txt', 'r') as y_data:
    y = y_data.read().split('\n')
    temp = [float(i) for i in y if i != '']
    y = temp

result = dft(x, y)

with open('./result.txt', 'a') as result_file:
    result_file.truncate(0)
    for i in result:
        result_file.write(f'{i[0]} {i[1]} {i[2]} \n')

print('Done!')
